core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+     * <ol>
+     *     <li>{@link RequestPrepared}</li>
+     *     <li>{@link RequestHeaderSent}</li>
+     *     <li>{@link RequestBytesSent}</li>
+     *     <li>{@link ResponseHeaderReceived}</li>
+     *     <li>{@link ResponseBytesReceived}</li>
+     *     <li>{@link ExecutionSuccess}</li>
+     * </ol>
+     * Failed transfer interface hierarchy:
+     * <ol>
+     *     <li>{@link RequestPrepared}</li>
+     *     <li>{@link AttemptFailure}</li>
+     *     <li>{@link ExecutionFailure}</li>
+     * </ol>
+     *
+     * @see ProgressListener
+     */
+    @SdkProtectedApi
+    final class Context {
+        private Context() {
+        }
+
+        /**
+         * A new transfer has been initiated.
+         * <p>
+         * Available context attributes:
+         * <ol>
+         *     <li>{@link RequestPrepared#request()}</li>
+         *     <li>{@link RequestPrepared#progressSnapshot()}</li>
+         * </ol>
+         */
+        @Immutable
+        @ThreadSafe
+        @SdkPublicApi
+        @SdkPreviewApi
+        public interface RequestPrepared {
+            /**
+             * The {@link SdkHttppRequest} that was submitted to SDK, i.e., the {@link PutObjectRequest} or
+             * {@link GetObjectRequest}.
+             */
+            SdkHttpRequest request();
+
+            /**
+             * The immutable {@link ProgressSnapahot} for this specific update.
+             */
+            ProgressSnapshot progressSnapshot();
+        }
+
+        /**
+         * The submitted {@link SdkHttppRequest} request header was successfully sent to the service
+         * <p>
+         * Available context attributes:
+         * <ol>
+         *     <li>{@link RequestHeaderSent#request()}</li>
+         *     <li>{@link RequestHeaderSent#progressSnapshot()}</li>
+         * </ol>
+         */
+        @Immutable
+        @ThreadSafe
+        @SdkPublicApi
+        @SdkPreviewApi
+        public interface RequestHeaderSent extends RequestPrepared {
+        }
+
+        /**
+         * Additional bytes sent
+         * <p>
+         * Available context attributes:
+         * <ol>
+         *     <li>{@link RequestBytesSent#request()}</li>
+         *     <li>{@link RequestBytesSent#progressSnapshot()}</li>
+         * </ol>
+         */
+        @Immutable
+        @ThreadSafe
+        @SdkPublicApi
+        @SdkPreviewApi
+        public interface RequestBytesSent extends RequestHeaderSent {
+        }
+
+        /**
+         * Service has sent back a response header, denoting the start of response reception
+         * <p>
+         * Available context attributes:
+         * <ol>
+         *     <li>{@link ResponseHeaderReceived#request()}</li>
+         *     <li>{@link ResponseHeaderReceived#progressSnapshot()}</li>
+         * </ol>
+         */
+        @Immutable
+        @ThreadSafe
+        @SdkPublicApi
+        @SdkPreviewApi
+        public interface ResponseHeaderReceived extends RequestBytesSent {
+        }
```

Comment:
Should we add the SdkHttpResponse here?

Reply:
Discussed offline. Added SdkHttpResponse